server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,218 @@
+     */
+    public @Nullable DiscoveryNode getMostRecentNonNullMaster() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        DiscoveryNode mostRecentNonNullMaster = null;
+        for (int i = masterHistoryCopy.size() - 1; i >= 0; i--) {
+            TimeAndMaster timeAndMaster = masterHistoryCopy.get(i);
+            if (timeAndMaster.master != null) {
+                mostRecentNonNullMaster = timeAndMaster.master;
+                break;
+            }
+        }
+        return mostRecentNonNullMaster;
+    }
+
+    /**
+     * Returns true if for the life of this MasterHistory (30 minutes) only one non-null node has been master, and the master has switched
+     * from that node to null n times.
+     * @param n The number of times the non-null master must have switched to null
+     * @return True if there has been a single non-null master and it has switched to null n or more times.
+     */
+    public boolean hasSameMasterGoneNullNTimes(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        return hasSameMasterGoneNullNTimes(
+            masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).collect(Collectors.toList()),
+            n
+        );
+    }
+
+    /**
+     * Returns true if for List of master nodes passed in only one non-null node has been master, and the master has switched
+     * from that node to null n times.
+     * @param masters The List of masters to use
+     * @param n The number of times the non-null master must have switched to null
+     * @return True if there has been a single non-null master and it has switched to null n or more times in the given list of masters.
+     */
+    public static boolean hasSameMasterGoneNullNTimes(List<DiscoveryNode> masters, int n) {
+        if (getDistinctMastersSeen(masters).size() != 1) {
+            return false;
+        }
+        boolean seenNonNull = false;
+        int timesMasterHasGoneNull = 0;
+        for (DiscoveryNode master : masters) {
+            if (master != null) {
+                seenNonNull = true;
+            } else if (seenNonNull) {
+                timesMasterHasGoneNull++;
+            }
+        }
+        return timesMasterHasGoneNull >= n;
+    }
+
+    private static Set<DiscoveryNode> getDistinctMastersSeen(List<DiscoveryNode> masters) {
+        return masters.stream().filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        return masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).filter(Objects::nonNull).collect(Collectors.toSet());
```

Comment:
Not quite, because that method takes `List<DiscoveryNode>` instead of `List<TimeAndMaster>`. Until yesterday it was pretty similar to what you suggest though -- https://github.com/elastic/elasticsearch/pull/85941#discussion_r859574627.